portal-impl/src/com/liferay/portal/template/TemplatePortletPreferences.java

```diff
@@ -34,29 +33,38 @@
 	public String getPreferences(Map<String, Object> preferences)
 		throws ReadOnlyException {
 
-		PortletPreferencesImpl portletPreferencesImpl =
-			new PortletPreferencesImpl();
+		StringBundler sb = new StringBundler();
+
+		sb.append("<portlet-preferences>");
 
 		for (Map.Entry<String, Object> entry : preferences.entrySet()) {
-			Object value = entry.getValue();
+			sb.append("<preference><name>");
+			sb.append(entry.getKey());
+			sb.append("</name>");
+
+			String[] values = _NULL_VALUE;
```

Comment:
It comes from the logic of PortletPreferencesImpl, in old logic, we invoke portletPreferencesImpl.setValue/setValues which will translate null object to "NULL_VALUE", though we did not really pass in a null from our ftl.

I am using Object as the value type because I would like to store both String and String[] in same map.